src/aiida/orm/comments.py

```diff
@@ -63,6 +64,47 @@ class Comment(entities.Entity['BackendComment', CommentCollection]):
 
     _CLS_COLLECTION = CommentCollection
 
+    __qb_fields__ = [
+        add_field(
+            'uuid',
+            dtype=str,
+            is_db_field=True,
+            doc='The UUID of the comment',
+        ),
+        add_field(
+            'ctime',
+            dtype=datetime,
+            is_db_field=True,
+            doc='Creation time of the comment',
+        ),
+        add_field(
+            'mtime',
+            dtype=datetime,
+            is_db_field=True,
+            doc='Modified time of the comment',
+        ),
+        add_field(
+            'content',
+            dtype=str,
+            is_db_field=True,
+            doc='Content of the comment',
+        ),
+        add_field(
+            'user_pk',
+            'user_id',
```

Comment:
Sure, I could address this in a future PR. You had mentioned the mapping being a bit hardcoded in this implementation, effectively destroying the backend abstraction. It's reasonable for the maps to move to the backend implementation classes.

Reply:
The future is now I guess? 😅

Do we agree on migrating the maps to the backend implementation classes?

Reply:
Which classes do you mean? Do you mean to change the sqlalchemy models themselves? Because that would require an actual database migration. I was more thinking to just have a mapping on the implementation in `aiida.storage.psql_dos.orm` classes, i.e. a "soft" mapping in the implementation, not actually changing the database schema.

Reply:
Before you do so, maybe you want to clean up your branch now to contain just 2 commits, the first by Chris and then a second with your fixes and additions. This way we can easily revert to the current state of the code since we know this is in a good state and pretty much ready to merge except from the aliases thing

Reply:
Sure, I clean first.

As for the first comment, not change the models, but rather move the mapping of `pk -> id` out to the implementation modules. I think we're saying the same thing if this is `aiida.storage.psql_dos.orm`.

Reply:
Cleaned! Let's fix a meeting to discuss the mapping change.

Reply:
Regarding soft mapping in `aiida.storage.psql_dos.orm`, how would you handle the case of `orm.TrajectoryData`, which is not a core SqlAlchemy type, that has the mapping:

``` python
__qb_fields__ = [
    add_field(
        'units_positions',
        'units|positions',
        dtype=str,
    ),
    add_field(
        'units_times',
        'units|times',
        dtype=str,
    ),
    add_field(
        'symbols',
        dtype=List[str],
        doc='list of symbols',
    ),
]
```

We seem to be using custom fields in the backend, e.g. `units|positions`. Is this a QE artifact? In any case, Chris's API allows this custom mapping at the front end. If we intend to do away with such an API and move it to the back end, how do you see such customization handled?

Reply:
Why do we need the version with the underscore? The class actually stores the attributes under the key with the `|` character. The model merely reflects this.

This is different from the `pk/id` case. In the front-end, we decided a long time ago that we prefer `pk` since `id` is a reserved keyword in Python. However, at that point the database models already existed and contain `id`. That is why both `pk` and `id` can often be used, and it some case only `id` can be used since essentially the user is directly interacting with the backend. This is the place where we need proper mapping if we want to use just `pk` in the frontend, but allow the backend to keep `id`.

Reply:
The first argument to `add_field` is actually the one used programmatically. Can't do `orm.TrajectoryData.fields.units|positions`.

Reply:
I see. The pipe character is also used by the `ArrayData` class, which is the parent of `TrajectoryData`. These were some of the first classes in AiiDA and were there before I joined, but looking at the code, I would guess that they chose to use the pipe for certain attribute keys in order to distinguish them from other "custom" attributes. This is used to be able to find the names of the arrays that are stored by the node.

There are two options that I see:

1. Change your code in order to not have to use the first argument passed to `add_field` programmatically. (Where exactly is this currently used? Would it be that bad not to use it programmatically? If so, we would also have to add validation/limitations on what are considered valid field identifiers. They essentially would have to be valid Python identifiers. We already enforce this as well for the labels of links between nodes).
2. Change the node implementations that use pipes in attributes, replacing them with underscores. This would require a data migration to adapt existing instances. This would minimize backwards incompatibility, but it will break existing code that accesses the attributes in their old form, e.g., `ArrayData().base.attributes.get('array|name')`.

Maybe there is a third, where we allow and keep pipes in attribute keys, and then we put the mapping somewhere (I have to think where this would be). But this would have to be done in a general way and cannot be specific to the node types we define in `aiida-core`, because then it would not work for any data types from plugin packages that also use incompatible characters in attribute keys.

Thinking about it that way, I don't think we currently impose any limitations on attribute keys (on the front end at least). Anything that can be stored in Postgres' JSONB is accepted. If we start to put limitations now, we would have to think about this closely how to deal with existing plugins that may violate it.

Reply:
Regarding 1, they are used programmatically in the query builder, in defining filters and projections. This is the new syntax introduced by Chris. I don't follow your comment on valid field identifiers.

Regarding 2, I'm not convinced this is necessary.

I think it is best we meet to discuss this. We may not be entirely on the same page.